Mathlib/Data/ENNReal/Floor.lean

```diff
@@ -0,0 +1,162 @@
+/-
+Copyright (c) 2024 Kalle Kytölä. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kalle Kytölä
+-/
+import Mathlib.Topology.Instances.ENNReal
+import Mathlib.Data.Real.ENatENNReal
+
+/-!
+# A floor function on the extended non-negative reals
+-/
+
+open Filter BigOperators TopologicalSpace Topology Set ENNReal NNReal ENat
+
+section ENat_topology
+
+instance : TopologicalSpace ℕ∞ := TopologicalSpace.induced ENat.toENNReal inferInstance
+
+lemma ENat.continuous_toENNReal : Continuous ENat.toENNReal :=
+  continuous_iff_le_induced.mpr fun _ h ↦ h
+
+instance : OrderTopology ℕ∞ := sorry
+
+-- TODO: Move to the appropriate file `Data.ENat.Basic`.
+lemma ENat.toENNReal_coe_eq_top_iff {m : ℕ∞} :
+    (m : ℝ≥0∞) = ∞ ↔ m = ⊤ := by
+  rw [← toENNReal_coe_eq_iff, toENNReal_top]
+
+-- TODO: Move to the appropriate file `Data.ENat.Basic`.
+lemma ENat.toENNReal_coe_ne_top_iff {m : ℕ∞} :
+    (m : ℝ≥0∞) ≠ ∞ ↔ m ≠ ⊤ :=
+  not_iff_not.mpr toENNReal_coe_eq_top_iff
+
+end ENat_topology
+
+section ENNReal_floor
+
+namespace ENNReal
+
+/-- The floor function `ℝ≥0∞ → ℕ∞`: the floor of `x` is the supremum of the extended natural
+numbers `n` satisfying `n ≤ x`. -/
+noncomputable def floor (x : ℝ≥0∞) : ℕ∞ := sSup {n : ℕ∞ | n ≤ x}
```

Comment:
I would tend to think this should be called `ENat.floor` to match `Nat.floor` and `Int.floor`. Other types it could apply to are `ENNRat` (doesn't exist yet), `EReal`, `ERat`

Reply:
Yes, that and mimicking the API of `Nat.floor` makes sense now if this is made `ENat`-valued. (The topology on `ENat` is my main concern with this currently. And to a smaller extent the convenience of the required casts back to `ENNReal`, since there does not seem to be a lot of API for `ENat`, and `ENNReal` by itself can already be less-than-pleasant to work with at times.)

I originally thought the `ENNReal.floor` would be a one-off and therefore different from the more standard floor designs, but maybe someone can some day have use for a floor function on the `ENNRat`, too.

I will try to make this more parallel to the existing one. Thanks for the comment and the improvements to code!

Reply:
The good thing about making your function be `ENat`-valued is that it suddenly forms a Galois coinsertion: `GaloisCoinsertion Int.cast FloorRing.floor` should hold.